src/ripple/unl/tests/BasicNetwork.h

```diff
@@ -392,124 +690,137 @@ BasicNetwork<Peer, Derived>::disconnect(
         links_[&from],
             link_transform{ *this, from });
 }
 
-template <class Peer, class Derived>
+template <class Peer>
+template <class Function>
+inline
+void
+BasicNetwork<Peer>::send(
+    Peer& from, Peer& to, Function&& f)
+{
+    using namespace std;
+    auto const iter =
+        links_[&from].find(&to);
+    queue_.emplace(&from, &to,
+        clock_.now() + iter->second.delay,
+            forward<Function>(f));
+}
+
+template <class Peer>
+template <class Function>
 inline
 auto
-BasicNetwork<Peer, Derived>::peers(
-    Peer& from) ->
-        boost::transformed_range<
-            peer_transform, links_type>
+BasicNetwork<Peer>::timer(
+    time_point const& when, Function&& f) ->
+        cancel_token
 {
-    return boost::adaptors::transform(
-        links_[&from], peer_transform{});
+    using namespace std;
+    return queue_.emplace(
+        nullptr, nullptr, when,
+            forward<Function>(f));
 }
 
-template <class Peer, class Derived>
-template <class Message>
+template <class Peer>
+template <class Function>
+inline
+auto
+BasicNetwork<Peer>::timer(
+    duration const& delay, Function&& f) ->
+        cancel_token
+{
+    return timer(clock_.now() + delay,
+        std::forward<Function>(f));
+}
+
+template <class Peer>
 inline
 void
-BasicNetwork<Peer, Derived>::send(
-    Peer& from, Peer& to, Message&& m)
+BasicNetwork<Peer>::cancel(
+    cancel_token const& token)
 {
-    auto const iter = links_[&from].find(&to);
-    msgs_.emplace(from, to, clock_.now() + iter->second.delay,
-        std::make_unique<invoke_impl<Message>>(
-            derived(), to, std::forward<Message>(m)));
+    queue_.erase(token.iter_);
 }
 
-template <class Peer, class Derived>
+template <class Peer>
 bool
-BasicNetwork<Peer, Derived>::run_one()
+BasicNetwork<Peer>::step_one()
 {
-    using boost::multi_index::get;
-    auto& by_when = get<2>(msgs_);
-    if (by_when.empty())
+    if (queue_.empty())
         return false;
-    auto const iter = by_when.begin();
-    auto const from = iter->from;
-    auto const op = std::move(iter->op);
+    auto const iter = queue_.begin();
     clock_.set(iter->when);
-    by_when.erase(iter);
-    (*op)(*from);
+    (*iter)();
+    queue_.erase(iter);
     return true;
 }
 
-template <class Peer, class Derived>
+template <class Peer>
 bool
-BasicNetwork<Peer, Derived>::run()
+BasicNetwork<Peer>::step()
 {
-    if (! run_one())
+    if (! step_one())
         return false;
     for(;;)
-        if (! run_one())
+        if (! step_one())
             break;
     return true;
 }
 
-template <class Peer, class Derived>
+template <class Peer>
 bool
-BasicNetwork<Peer, Derived>::run_until(
+BasicNetwork<Peer>::step_until(
     time_point const& until)
 {
-    using boost::multi_index::get;
-    auto& by_when = get<2>(msgs_);
-    if(by_when.empty() ||
-        by_when.begin()->when > until)
+    // VFALCO This routine needs optimize
+    if(queue_.empty())
+        return false;
```

Comment:
You're right, its incorrect. The clock should be advanced to `until`. I noticed this and fixed it in my continued work here:
https://github.com/vinniefalco/rippled/blob/sim2/src/ripple/unl/tests/BasicNetwork.h#L798

Its probably worth fixing in this PR.